src/surface/poisson_disk_sampler.cpp

```diff
@@ -0,0 +1,270 @@
+  activeList.push_back(sample);
+  addPointToSpatialLookup(sample.interpolate(geometry.vertexPositions));
+}
+
+// ===== Helpers for spatial lookup. This code was pretty much directly taken from Nick Sharp's Poisson sampling code.
+
+PoissonDiskSampler::SpatialKey PoissonDiskSampler::positionKey(const Vector3& position) const {
+
+  // Each bucket has sidelength r/(sqrt(3) * 2) so that it can hold at most 1 sample, and any two points within two
+  // adjacent buckets are guaranteed to be within r of each other.
+  Vector3 coord = (position - mapCenter) / r * (2.0 * std::sqrt(3.0));
+  long long int keyX = static_cast<long long int>(std::floor(coord.x));
+  long long int keyY = static_cast<long long int>(std::floor(coord.y));
+  long long int keyZ = static_cast<long long int>(std::floor(coord.z));
+
+  return {keyX, keyY, keyZ};
+}
+
+/*
+ * Optionally mark all buckets with a radius of <R> buckets as occupied, as well.
+ */
+void PoissonDiskSampler::addPointToSpatialLookup(const Vector3& newPoint, int R) {
+
+  SpatialKey key = positionKey(newPoint);
+
+  spatialBuckets.insert(std::make_pair(key, true));
+
+  if (R < 0) return;
+
+  // implicitly convert to floats
+  long long int keyX = key[0];
+  long long int keyY = key[1];
+  long long int keyZ = key[2];
+  long long int queryX, queryY, queryZ;
+  for (int offsetX = -R; offsetX <= R; offsetX++) {
+    for (int offsetY = -R; offsetY <= R; offsetY++) {
+      for (int offsetZ = -R; offsetZ <= R; offsetZ++) {
+
+        queryX = keyX + offsetX;
+        queryY = keyY + offsetY;
+        queryZ = keyZ + offsetZ;
+
+        long long int dx = queryX - keyX;
+        long long int dy = queryY - keyY;
+        long long int dz = queryZ - keyZ;
+        if (std::sqrt(dx * dx + dy * dy + dz * dz) < (float)R) {
+          SpatialKey newKey = {queryX, queryY, queryZ};
+          spatialBuckets.insert(std::make_pair(newKey, true));
+        }
+      }
+    }
+  }
+}
+
+/*
+ * Given a candidate point, return "false" if the smallest distance to any of the existing samples is less than r.
+ * Return "true" otherwise.
+ */
+bool PoissonDiskSampler::isCandidateValid(const SurfacePoint& candidate) const {
+
+  Vector3 queryPoint = candidate.interpolate(geometry.vertexPositions);
+  SpatialKey queryKey = positionKey(queryPoint);
+
+  // Check all adjacent buckets. Each bucket's sidelength is chosen to be small enough such that if any two adjacent
+  // buckets contain points, their points are guaranteed to be within r of each other.
+  for (int offsetX = -1; offsetX <= 1; offsetX++) {
+    for (int offsetY = -1; offsetY <= 1; offsetY++) {
+      for (int offsetZ = -1; offsetZ <= 1; offsetZ++) {
+
+        SpatialKey bucketKey = queryKey;
+        bucketKey[0] += offsetX;
+        bucketKey[1] += offsetY;
+        bucketKey[2] += offsetZ;
+
+        auto bucketIter = spatialBuckets.find(bucketKey);
+        if (bucketIter != spatialBuckets.end()) {
+          if (bucketIter->second) {
```

Comment:
Is there a reason that this just checks whether the bucket is occupied, rather than storing the actual points in the buckets and checking the distance to those points?

I think that simply checking the buckets for non-emptyness could potentially lead to axis-aligned sampling artifacts, right? Because the spherical neighborhood of points within distance `r` does not exactly overlap with the cubic buckets. 

However we could instead choose the bucket size such that the 3x3x3 grid of buckets contains the sphere, then store the actual points in the buckets, and check the distance to the point inside of a any nonempty buckets before returning `false`.

Although I can imagine that any sampling bias that this strategy introduces is not a visible problem in practice.

Reply:
Yes, you're correct. In my haste to save an extra computation, I failed to recognize that (two points in adjacent buckets are guaranteed to be within r of each other) does not imply (two points that are not in adjacent buckets are guaranteed to be greater than r away from each other.)

I've followed your suggestion, and set the side length of the buckets so that each bucket contains at most one point, and store the actual points in the buckets.